Group methods in help() by class.

adorokhine opened the pull request:
Example:
```
>>> print s.help()

Methods in class EventSnippet:
  eventGetAll(String, String) returns List  // Gets and removes all the events of a certain name that have been received so far. Non-blocking. Potentially racey since it does not guarantee no event of the same name will occur after the call.
  eventWaitAndGet(String, String, Integer) returns JSONObject  // Blocks until an event of a specified type has been received. The returned event is removed from the cache. Default timeout is 60s.

Methods in class ExampleSnippet:
  getFoo(Integer) returns String  // Returns the given integer with the prefix "foo"

Methods in class ExampleSnippet2:
  getBar(String) returns String  // Returns the given string with the prefix "bar"
  throwSomething() returns String  // Throws an exception
  throwSomethingFromMainThread() returns String  // Throws an exception from the main thread

```
Fixes #44.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/mobly-snippet-lib/57)
<!-- Reviewable:end -->

xpconanfan commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-KjL1pxx9d_zFhq51sYP:bgog0yr) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
> ```Java
>         StringBuilder result = new StringBuilder();
>         for (Map.Entry<String, Set<MethodDescriptor>> entry : methods.entrySet()) {
>             result.append("\nMethods in class ").append(entry.getKey()).append(":\n");
> ```

"Methods in class" is not really accurate since we are only supposed to list public methods annotated with `Rpc` here.
How about "Rpc in XxxSnippet", or simply use the class name?

Another minor concern is, with this extra layer of text in help menu, people might read it as, "I need to use the class name when calling the methods". Thoughts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-KjL8yP2gaPn7NU7OWgj:b-qjhwsn) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

"Methods in class" is not really accurate since we are only supposed to list public methods annotated with `Rpc` here.
How about "Rpc in XxxSnippet", or simply use the class name?

Another minor concern is, with this extra layer of text in help menu, people might read it as, "I need to use the class name when calling the methods". Thoughts?
</blockquote></details>

I had a similar impression/concern when using just the class name, which is why I added the extra text. Let me know what you'd like to see instead. Changed to Rpc Methods in Classname for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-KjLDPySEOQW3eoBGDPK:bsmopax) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

I had a similar impression/concern when using just the class name, which is why I added the extra text. Let me know what you'd like to see instead. Changed to Rpc Methods in Classname for now.
</blockquote></details>

Maybe we can give an example on usage in the help menu?
Like
```
Call any of the following Rpc methods directly, like:
snippet_client.wifiEnable()
```

Also maybe we should mark out the ones that use `AsyncRpc`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-KjerEA1zhJQVujxL-Tw:bj17rza) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
> Maybe we can give an example on usage in the help menu?
Do you want this to be printed on the java side, or on the python side?

> Also maybe we should mark out the ones that use AsyncRpc?
Can you give me an example of what you want this to look like?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-Kjerblnw8Ei_9-4m3Tx:b-p7wvgr) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

> Maybe we can give an example on usage in the help menu?
Do you want this to be printed on the java side, or on the python side?

> Also maybe we should mark out the ones that use AsyncRpc?
Can you give me an example of what you want this to look like?
</blockquote></details>

1. I mean as part of help() so ppl know they don't need to worry about the extra layer of snippet class grouping when calling the Rpcs.
2. I'm not sure what the best way to mark AsyncRpc is.
Maybe:
```
XxxSnippet:
  [Async] doSomethingAsync(String, String) returns List  // Does something Asynchronously.
```
what do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-KjeyPH68JiWqivo7v3a:b986f6n) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
> Call any of the following Rpc methods directly, like: `snippet_client.wifiEnable()`

My concern is that if someone were to copy and paste this example invocation, they'd get an error like `name 'snippet_client' is not defined`.

I'd be happier printing it on the python side instead so we could give a real example, like `snippet.eventWaitAndGet()`. That would actually work if you copy and pasted it.

However the python side doesn't know it's running `help` versus any other rpc. I could print it by default instead of `Call methods against:` ?

Alternatively I could use 'snippet' on the java side with the implicit understanding that it should to match what snippet_shell binds it to on the python side?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-Kjf_6F4TPLYDWeR2icv:bfj0zs6) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

> Call any of the following Rpc methods directly, like: `snippet_client.wifiEnable()`

My concern is that if someone were to copy and paste this example invocation, they'd get an error like `name 'snippet_client' is not defined`.

I'd be happier printing it on the python side instead so we could give a real example, like `snippet.eventWaitAndGet()`. That would actually work if you copy and pasted it.

However the python side doesn't know it's running `help` versus any other rpc. I could print it by default instead of `Call methods against:` ?

Alternatively I could use 'snippet' on the java side with the implicit understanding that it should to match what snippet_shell binds it to on the python side?
</blockquote></details>

Instead, we could add a `help()` function in the Python client that adds extra lines to what the actual `help()` rpc call returns, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

dthkao commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-Kjh4AlrG2CfWZHPFJhK:bayhkuf) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
<details><summary><i>Previously, xpconanfan (Ang Li) wrote…</i></summary><blockquote>

Instead, we could add a `help()` function in the Python client that adds extra lines to what the actual `help()` rpc call returns, right?
</blockquote></details>

Why not just "Rpcs (provided|implemented) by XxxSnippet"? You can add the extra python client prefix to them if still needed, but I think from the java-only help() call this is pretty clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

adorokhine commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java, line 145 at r1](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-KjL1pxx9d_zFhq51sYO:-Kl1QL4xMCeTVFXOXP2N:b-896fix) ([raw file](https://github.com/google/mobly-snippet-lib/blob/6369d08184846cbf080a786786a0211561cc13b3/third_party/sl4a/src/main/java/com/google/android/mobly/snippet/rpc/JsonRpcServer.java#L145)):*
<details><summary><i>Previously, dthkao (David T.H. Kao) wrote…</i></summary><blockquote>

Why not just "Rpcs (provided|implemented) by XxxSnippet"? You can add the extra python client prefix to them if still needed, but I think from the java-only help() call this is pretty clear.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57)*
<!-- Sent from Reviewable.io -->

xpconanfan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly-snippet-lib/57#-:-Kl1QWDzT4B8Jq15vLe5:bnfp4nl)*
<!-- Sent from Reviewable.io -->
